Source/CarthageKit/Project.swift

```diff
@@ -571,26 +571,54 @@ public final class Project {
+				func filter(dependencies: Set<ProjectIdentifier>, by submodules: [Submodule]) -> [ProjectIdentifier] {
+					return dependencies.filter { identifier in
+						!(submodules.filter { (submodule: Submodule) -> Bool in
+							submodule.path.characters.prefix(9).last == .Some("/") && String(
+								submodule.path.characters.prefix(CarthageProjectCheckoutsPath.characters.count)
+							).caseInsensitiveCompare(CarthageProjectCheckoutsPath) == .OrderedSame
+						}.map { $0.name } as [String]).contains(identifier.name)
+					}
 				}
 				
-				if let submodule = submodule {
-					return addSubmoduleToRepository(self.directoryURL, submodule, GitURL(repositoryURL.path!))
+				/// The submodule for an already existing submodule at dependency project’s path
+				/// or the submodule to be added at this path given the `--use-submodules` flag.
+				func submodule() -> Submodule? {
+					if var foundSubmodule = submodulesByPath[project.relativePath] {
+						foundSubmodule.URL = repositoryURLForProject(project, preferHTTPS: self.preferHTTPS)
+						foundSubmodule.SHA = revision
+						return foundSubmodule
+					} else if self.useSubmodules {
+						return Submodule(name: project.relativePath, path: project.relativePath, URL: repositoryURLForProject(project, preferHTTPS: self.preferHTTPS), SHA: revision)
+					} else {
+						return nil
+					}
+				}
+				
+				let signal: SignalProducer<Submodule, CarthageError>
+				
+				if let submodule = submodule() {
+					// Submodules for subdependencies of `project` are recursed through and initialized by git.
+					signal = addSubmoduleToRepository(self.directoryURL, submodule, GitURL(repositoryURL.path!))
 						.startOnQueue(self.gitOperationQueue)
+						.then(submodulesInRepository(self.directoryURL, revision: revision))
 				} else {
-					return checkoutRepositoryToDirectory(repositoryURL, workingDirectoryURL, revision: revision)
-						.then(self.dependenciesForDependency(dependency))
-						.flatMap(.Merge) { dependencies in
-							return self.symlinkCheckoutPathsForDependencyProject(dependency.project, subDependencies: dependencies, rootDirectoryURL: self.directoryURL)
+					signal = checkoutRepositoryToDirectory(repositoryURL, workingDirectoryURL, revision: revision)
+						.then(submodulesInRepository(self.directoryURL, revision: revision))
+						.attempt /* just for side effects */ {
+							// For checkouts of “ideally bare” depositories, we clone submodules for subdependencies ourselves.
+							cloneSubmoduleInWorkingDirectory($0, workingDirectoryURL).single()!
 						}
 				}
+				
+				return signal.collect().flatMap(.Merge) { submodules in
+					// Symlink dependencies which do not match submodules.
+					self.dependenciesForDependency(dependency).flatMap(.Merge) { dependencies in
+						self.symlinkCheckoutPathsForDependencyProject(dependency.project, subDependencies: filter(dependencies, by: submodules), rootDirectoryURL: self.directoryURL)
+					}
+				}.then(.empty)
```

Comment:
This should use our standard formatting, where each operator is indented on its own line.

```
return signal
    .collect()
    .flatMap(.Merge) { submodules in
        …
    }
    .then(.empty)
```